### ISSUE TITLE
Dtype conversion error using numpy 1.21.0 release candidate

### ISSUE THREAD

**Issue by adeak** [HypothesisWorks/hypothesis#2975]
We have two tests [in PyVista](https://docs.pyvista.org/) that raise an error from within hypothesis using the new numpy release candidate. [One of the tests](https://github.com/pyvista/pyvista/blob/master/tests/test_datasetattributes.py#L150-L154):
```py
@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
@given(arr=arrays(dtype='U', shape=10))
def test_append_string_array_should_equal(arr, hexbeam_field_attributes):
    hexbeam_field_attributes['string_arr'] = arr
    assert np.array_equiv(arr, hexbeam_field_attributes['string_arr'])
```

The error appears [here](https://github.com/HypothesisWorks/hypothesis/blob/master/hypothesis-python/src/hypothesis/extra/numpy.py#L301) and it has to do with trying to convert from object dtype to string which almost always works... except for one case that is somehow generated by hypothesis:
```py
>>> import numpy as np
>>> np.array([''], dtype=object).astype('<U0')  # or just .astype('U')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Cannot cast array data from dtype('O') to dtype('<U') according to the rule 'unsafe'
```

This worked on numpy 1.20.3. I don't understand how hypothesis works, but this seems to be something that needs fixed to stay compatible with the next numpy release. Or is our test using hypothesis wrong?

In the meantime I'll also ask on the numpy side whether this change is expected ([link](https://github.com/numpy/numpy/issues/19085)).

Full test traceback, although direct reproducibility would need installing pyvista:
<details>
  <summary>Click for output</summary>

  ```py
  ==================================================================================================== FAILURES =====================================================================================================
  ______________________________________________________________________________________ test_append_string_array_should_equal ______________________________________________________________________________________
  
  hexbeam_field_attributes = pyvista DataSetAttributes
  Association: NONE
  Contains keys:
  	
  
  >   ???
  
  tests/test_datasetattributes.py:151: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/core.py:709: in _execute_once_for_engine
      escalate_hypothesis_internal_error()
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/core.py:683: in _execute_once_for_engine
      result = self.execute_once(data)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/core.py:619: in execute_once
      result = self.test_runner(data, run)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/executors.py:52: in default_new_style_executor
      return function(data)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/core.py:578: in run
      args, kwargs = data.draw(self.search_strategy)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/internal/conjecture/data.py:888: in draw
      return strategy.do_draw(self)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/core.py:223: in do_draw
      return self.mapped_strategy.do_draw(data)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/strategies/_internal/collections.py:57: in do_draw
      return tuple(data.draw(e) for e in self.element_strategies)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/strategies/_internal/collections.py:57: in <genexpr>
      return tuple(data.draw(e) for e in self.element_strategies)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
      return strategy.do_draw(self)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py:801: in do_draw
      result = self.pack(data.draw(self.mapped_strategy))
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
      return strategy.do_draw(self)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py:168: in do_draw
      return data.draw(self.wrapped_strategy)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
      return strategy.do_draw(self)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py:801: in do_draw
      result = self.pack(data.draw(self.mapped_strategy))
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
      return strategy.do_draw(self)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/strategies/_internal/collections.py:57: in do_draw
      return tuple(data.draw(e) for e in self.element_strategies)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/strategies/_internal/collections.py:57: in <genexpr>
      return tuple(data.draw(e) for e in self.element_strategies)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
      return strategy.do_draw(self)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py:168: in do_draw
      return data.draw(self.wrapped_strategy)
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
      return strategy.do_draw(self)
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  
  self = <hypothesis.extra.numpy.ArrayStrategy object at 0x7fc15470a760>, data = ConjectureData(VALID, 2 bytes, frozen)
  
      def do_draw(self, data):
          if 0 in self.shape:
              return np.zeros(dtype=self.dtype, shape=self.shape)
      
          # Because Numpy allocates memory for strings at array creation, if we have
          # an unsized string dtype we'll fill an object array and then cast it back.
          unsized_string_dtype = (
              self.dtype.kind in ("S", "a", "U") and self.dtype.itemsize == 0
          )
      
          # This could legitimately be a np.empty, but the performance gains for
          # that would be so marginal that there's really not much point risking
          # undefined behaviour shenanigans.
          result = np.zeros(
              shape=self.array_size, dtype=object if unsized_string_dtype else self.dtype
          )
      
          if self.fill.is_empty:
              # We have no fill value (either because the user explicitly
              # disabled it or because the default behaviour was used and our
              # elements strategy does not produce reusable values), so we must
              # generate a fully dense array with a freshly drawn value for each
              # entry.
              if self.unique:
                  seen = set()
                  elements = cu.many(
                      data,
                      min_size=self.array_size,
                      max_size=self.array_size,
                      average_size=self.array_size,
                  )
                  i = 0
                  while elements.more():
                      # We assign first because this means we check for
                      # uniqueness after numpy has converted it to the relevant
                      # type for us. Because we don't increment the counter on
                      # a duplicate we will overwrite it on the next draw.
                      self.set_element(data, result, i)
                      if result[i] not in seen:
                          seen.add(result[i])
                          i += 1
                      else:
                          elements.reject()
              else:
                  for i in range(len(result)):
                      self.set_element(data, result, i)
          else:
              # We draw numpy arrays as "sparse with an offset". We draw a
              # collection of index assignments within the array and assign
              # fresh values from our elements strategy to those indices. If at
              # the end we have not assigned every element then we draw a single
              # value from our fill strategy and use that to populate the
              # remaining positions with that strategy.
      
              elements = cu.many(
                  data,
                  min_size=0,
                  max_size=self.array_size,
                  # sqrt isn't chosen for any particularly principled reason. It
                  # just grows reasonably quickly but sublinearly, and for small
                  # arrays it represents a decent fraction of the array size.
                  average_size=math.sqrt(self.array_size),
              )
      
              needs_fill = np.full(self.array_size, True)
              seen = set()
      
              while elements.more():
                  i = cu.integer_range(data, 0, self.array_size - 1)
                  if not needs_fill[i]:
                      elements.reject()
                      continue
                  self.set_element(data, result, i)
                  if self.unique:
                      if result[i] in seen:
                          elements.reject()
                          continue
                      else:
                          seen.add(result[i])
                  needs_fill[i] = False
              if needs_fill.any():
                  # We didn't fill all of the indices in the early loop, so we
                  # put a fill value into the rest.
      
                  # We have to do this hilarious little song and dance to work
                  # around numpy's special handling of iterable values. If the
                  # value here were e.g. a tuple then neither array creation
                  # nor putmask would do the right thing. But by creating an
                  # array of size one and then assigning the fill value as a
                  # single element, we both get an array with the right value in
                  # it and putmask will do the right thing by repeating the
                  # values of the array across the mask.
                  one_element = np.zeros(
                      shape=1, dtype=object if unsized_string_dtype else self.dtype
                  )
                  self.set_element(data, one_element, 0, self.fill)
                  if unsized_string_dtype:
  >                   one_element = one_element.astype(self.dtype)
  E                   TypeError: Cannot cast array data from dtype('O') to dtype('<U') according to the rule 'unsafe'
  
  ../tmp_test_env/lib/python3.8/site-packages/hypothesis/extra/numpy.py:301: TypeError
  --------------------------------------------------------------------------------------------------- Hypothesis ----------------------------------------------------------------------------------------------------
  You can add @seed(49582184409187967324836935015787832203) to this test or run pytest with --hypothesis-seed=49582184409187967324836935015787832203 to reproduce this failure.

  ```
</details>

**Comment by Zalathar** [comment]
At a glance, it looks like Hypothesis is doing some shenanigans to work around a numpy quirk, and those shenanigans are no longer allowed (by default?) in numpy 1.21.0.

So PyVista is probably innocent here.

**Comment by adeak** [comment]
Thanks for the feedback, @Zalathar. In the meantime I got a response on the numpy issue that suggests that this might be fixed in numpy before 1.21 is released, which would mean that neither hypothesis would have to fix anything :) (Shenanigans not withstanding.)

**Comment by Zac-HD** [comment]
Thanks very much for a great bug report @adeak!  

Per https://github.com/numpy/numpy/issues/19085#issuecomment-847406115 it looks like this is a regression in Numpy, so I don't think there's anything for us to do in Hypothesis.  If a stable Numpy version is released without fixing this, we will of course patch Hypothesis to work around it then 🙂
